[Lockdown Mode] Disable XSLT parsing for WebKit in Lockdown Mode

ronan-apple opened the pull request:
#### 261b24ff1774bf055cd4378f1bd12ca926417b35
<pre>
[Lockdown Mode] Disable XSLT parsing for WebKit in Lockdown Mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=295107">https://bugs.webkit.org/show_bug.cgi?id=295107</a>
&lt;<a href="https://rdar.apple.com/problem/151845594">rdar://problem/151845594</a>&gt;

Reviewed by Ryosuke Niwa.

This patch disables XSLT support when in Lockdown Mode by removing
access to the XSLTProcessor constructor to prevent JavaScript usage,
and preventing detection of a stylesheet as XSL unless a newly
introduced feature flag is enabled.

The feature flag is enabled by default and disabled in LDM to prevent
any change in current behaviour.

Added LayoutTests that ensure XSLTProcessor is inaccessible, and suitably
check an XML document with a stylesheet is not processed when in LDM.

* LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled-expected.txt: Added.
* LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled.html: Added.
* LayoutTests/dom/xsl/lockdown-mode/resources/XSLT-disabled.xml: Added.
* LayoutTests/dom/xsl/lockdown-mode/resources/XSLT-disabled.xsl: Added.
* LayoutTests/js/dom/lockdown-mode/XSLTProcessor-disabled-expected.txt: Added.
* LayoutTests/js/dom/lockdown-mode/XSLTProcessor-disabled.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::scheduleToApplyXSLTransforms):
(WebCore::Document::applyPendingXSLTransformsTimerFired):
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::checkStyleSheet):
* Source/WebCore/xml/XSLTProcessor.idl:

Canonical link: <a href="https://commits.webkit.org/297851@main">https://commits.webkit.org/297851@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/28df4ec56d66f0e8be1364900b7f130f862a83e4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/113069 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/32804 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/23282 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/119273 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/63663 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/115031 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/33456 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/41367 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/86058 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/36726 "Build was cancelled. Recent messages:Printed configuration") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/116016 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/26714 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/101712 "Passed tests") | [❌ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/66368 "Found 139 new API test failures: /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/configuration, /WPE/TestBackForwardList:/webkit/WebKitWebView/navigation-after-session-restore, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/reset, /WPE/TestInputMethodContext:/webkit/WebKitInputMethodContext/surrounding, /WPE/TestLoaderClient:/webkit/WebKitURIResponse/http-headers, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone, /WPE/TestWebKitWebContext:/webkit/WebKitWebContext/timezone-worker, /WPE/TestWebKitUserContentManager:/webkit/WebKitUserContentManager/injected-script, /WPE/TestLoaderClient:/webkit/WebKitWebView/unfinished-subresource-load, /WPE/TestWebsiteData:/webkit/WebKitWebsiteData/cookies ... (failure)") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/25992 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/19843 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/63031 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/105586 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/96101 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/19916 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/122494 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/111685 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/40147 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/29941 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/94909 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/40531 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/97931 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/94651 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24164 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/39792 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/17601 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/36272 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/40033 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/45532 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/135915 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/39674 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/36490 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/43007 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/41411 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

lilyspiniolas commented on LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled.xsl:
```diff
@@ -0,0 +1,10 @@
+    Transform - <xsl:value-of select="."/>
+  </xsl:template>
+
+</xsl:stylesheet>
```
nit: add empty newline at EOF

ronan-apple replied:
Ah, thank you! Fixed!

lilyspiniolas commented on LayoutTests/js/dom/lockdown-mode/script-tests/XSLTProcessor-disabled.js:
```diff
@@ -0,0 +1,5 @@
+"XSLTProcessor disabled test"
+);
+
+shouldBeTrue("typeof(XSLTProcessor) === 'undefined'");
```
nit: add empty newline at EOF

ronan-apple replied:
Ah, thank you! Fixed!

nt1m commented on LayoutTests/platform/ios/dom/xsl/lockdown-mode/XSLT-disabled-expected.txt:
Can we avoid the render tree dump?

ronan-apple replied:
Good question! There are two paths we want to test for the XSLT disablement - the first is the JS APIs being exposed, and the second is ensuring that browsing to an XML page doesn't trigger XSLT processing. 

This is for the latter - I couldn't find a way to do differently, as we can't use JS or similar to test whether processing occurred. If there's some other way of testing the page contents I'd appreciate learning how to. for sure!

ronan-apple replied:
Scratch that, I think I have an option to handle this by loading the XML in an iframe and then checking its contents. Will post the fix shortly!

ronan-apple replied:
Change pushed for this now - thanks for the suggestion!

nt1m commented on LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled.html:
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
+<html>
+<head>
+<script src="../../../resources/js-test-pre.js"></script>
```
```suggestion
<script src="../../../resources/js-test.js"></script>
```

The pre/post variants are obsolete.

ronan-apple replied:
Ah, ace, great to know - thank you! Pushed new change with this corrected.

nt1m commented on LayoutTests/js/dom/lockdown-mode/XSLTProcessor-disabled.html:
```diff
@@ -0,0 +1,10 @@
+</head>
+<body>
+<script src="script-tests/XSLTProcessor-disabled.js"></script>
+<script src="../../../resources/js-test-post.js"></script>
```
ditto here

ronan-apple replied:
Ace, thanks! Pushed with the new fix.

rniwa commented on Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
```diff
@@ -9371,6 +9371,21 @@ WritingSuggestionsAttributeEnabled:
     WebCore:
       default: true
 
+XsltEnabled:
```
Nit: We should capitalize Xslt -> XSLT.

ronan-apple replied:
Thanks, fixed! (And added fields in the preferences file to ensure the derived sources use reasonable names too).

rniwa commented on LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled.html:
```diff
@@ -0,0 +1,25 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
```
Use modern HTML5 style doc type: `<!DOCTYPE html>`

ronan-apple replied:
Thanks, fixed!

rniwa commented on LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled.html:
```diff
@@ -0,0 +1,25 @@
+
+    xmlDocument = document.getElementById("xml-iframe").contentDocument;
+
+    shouldBe("xmlDocument.children.length", "1");
```
Did you verify that this test fails without your PR?

ronan-apple replied:
Yep - this first check passes, but the rest of the checks all fail due to the transformed elements.

rniwa commented on LayoutTests/js/dom/lockdown-mode/XSLTProcessor-disabled.html:
```diff
@@ -0,0 +1,9 @@
+<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML//EN">
```
Use modern doctype: `<!DOCTYPE html>`.

ronan-apple replied:
Thanks, fixed!

rniwa commented on LayoutTests/js/dom/lockdown-mode/XSLTProcessor-disabled.html:
```diff
@@ -0,0 +1,9 @@
+<script src="../../../resources/js-test.js"></script>
+</head>
+<body>
+<script src="script-tests/XSLTProcessor-disabled.js"></script>
```
Don't split the test code into a separate js file. It makes harder to read the test code. Just directly embed it here.

ronan-apple replied:
Makes sense, thanks! Have moved this into this same file now.

rniwa commented on Source/WebCore/dom/Document.cpp:
```diff
@@ -7718,6 +7718,8 @@ bool Document::shouldDeferAsynchronousScriptsUntilParsingFinishes() const
 
 void Document::scheduleToApplyXSLTransforms()
 {
+    ASSERT(settings().xsltEnabled());
```
Can we add the same assertion to `applyPendingXSLTransformsTimerFired`?

ronan-apple replied:
Yep, absolutely, I've added it there too now, thanks!

Commit:
[Lockdown Mode] Disable XSLT parsing for WebKit in Lockdown Mode
https://bugs.webkit.org/show_bug.cgi?id=295107
<rdar://problem/151845594>

Reviewed by Ryosuke Niwa.

This patch disables XSLT support when in Lockdown Mode by removing
access to the XSLTProcessor constructor to prevent JavaScript usage,
and preventing detection of a stylesheet as XSL unless a newly
introduced feature flag is enabled.

The feature flag is enabled by default and disabled in LDM to prevent
any change in current behaviour.

Added LayoutTests that ensure XSLTProcessor is inaccessible, and suitably
check an XML document with a stylesheet is not processed when in LDM.

* LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled-expected.txt: Added.
* LayoutTests/dom/xsl/lockdown-mode/XSLT-disabled.html: Added.
* LayoutTests/dom/xsl/lockdown-mode/resources/XSLT-disabled.xml: Added.
* LayoutTests/dom/xsl/lockdown-mode/resources/XSLT-disabled.xsl: Added.
* LayoutTests/js/dom/lockdown-mode/XSLTProcessor-disabled-expected.txt: Added.
* LayoutTests/js/dom/lockdown-mode/XSLTProcessor-disabled.html: Added.
* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* LayoutTests/platform/win/TestExpectations:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::scheduleToApplyXSLTransforms):
(WebCore::Document::applyPendingXSLTransformsTimerFired):
* Source/WebCore/dom/ProcessingInstruction.cpp:
(WebCore::ProcessingInstruction::checkStyleSheet):
* Source/WebCore/xml/XSLTProcessor.idl:

Canonical link: https://commits.webkit.org/297851@main
